### PR TITLE
[selenium manager]: fix edge artifact deserialisation

### LINKED ISSUE #14851: [🐛 Bug]: Selenium Manager unable to lookup Edge - Error parsing JSON
### What happened?

Running Selenium Manager Windows 0.4.27 command line to pull any Edge version shows:
```
C:\AFT>selenium-manager-windows.exe --version
selenium-manager 0.4.27

C:\AFT>selenium-manager-windows.exe --browser edge --browser-version stable --trace --force-browser-download
[2024-12-04T09:28:52.541Z TRACE] Reading metadata from C:\Users\aftadmin\.cache\selenium\se-metadata.json
[2024-12-04T09:28:52.544Z TRACE] Metadata file does not exist. Creating a new one
[2024-12-04T09:28:52.545Z DEBUG] Sending stats to Plausible: Props { browser: "edge", browser_version: "stable", os: "windows", arch: "amd64", lang: "", selenium_version: "4.27" }
[2024-12-04T09:28:52.545Z TRACE] Writing metadata to C:\Users\aftadmin\.cache\selenium\se-metadata.json
[2024-12-04T09:28:52.546Z TRACE] Checking msedgedriver in PATH
[2024-12-04T09:28:52.573Z DEBUG] msedgedriver not found in PATH
[2024-12-04T09:28:52.574Z TRACE] Reading metadata from C:\Users\aftadmin\.cache\selenium\se-metadata.json
[2024-12-04T09:28:52.575Z DEBUG] Checking edge releases on https://edgeupdates.microsoft.com/api/products/
[2024-12-04T09:28:52.623Z ERROR] Error parsing JSON from URL https://edgeupdates.microsoft.com/api/products/ missing field `ArtifactName` at line 1 column 517
```

When I look at https://edgeupdates.microsoft.com/api/products/ the json has an "artifactName" (lower case a at start), and only for the browser, not the webdriver - I'm not sure of the format SM expects?
image

### How can we reproduce the issue?

```shell
See above.
```

### Relevant log output

```shell
See above.
```

### Operating System

Windows 10

### Selenium version

selenium-manager 0.4.27

### What are the browser(s) and version(s) where you see this issue?

Edge all versions

### What are the browser driver(s) and version(s) where you see this issue?

Edge all versions

### Are you using Selenium Grid?

n/a

### PR DESCRIPTION
**Pull request by Delta456**
### **User description**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description

Fixes https://github.com/SeleniumHQ/selenium/issues/14851

### Motivation and Context

The latest API of Microsoft Edge product updates now has renamed some fields which causes a lot of tests and selenium manager to fail.

API: https://edgeupdates.microsoft.com/api/products/

Previous:

```json
       "Artifacts": [
          {
            "ArtifactName": "msi",
            "Location": "https://msedge.sf.dl.delivery.mp.microsoft.com/filestreamingservice/files/91731797-32c8-494b-ba04-1083d8ebaf73/MicrosoftEdgeEnterpriseX64.msi",
            "Hash": "15F1ECC09251FF7D516B5B1BD4B48E07434A0546C6FB8822BEED1026DB5D4975",
            "HashAlgorithm": "SHA256",
            "SizeInBytes": 182636544
          }
 ``` 

Now:

```json
       "Artifacts": [
          {
            "artifactName": "msi",
            "location": "https://msedge.sf.dl.delivery.mp.microsoft.com/filestreamingservice/files/91731797-32c8-494b-ba04-1083d8ebaf73/MicrosoftEdgeEnterpriseX64.msi",
            "hash": "15F1ECC09251FF7D516B5B1BD4B48E07434A0546C6FB8822BEED1026DB5D4975",
            "hashAlgorithm": "SHA256",
            "sizeInBytes": 182636544
          }
 ``` 

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

___

### **PR Type**
Bug fix

___

### **Description**
- Fixed deserialization issues in the Selenium Manager for Microsoft Edge by updating field names in the `Artifact` struct to match the latest API response from Microsoft.
- This change addresses the JSON parsing error due to renamed fields in the Edge product updates API.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>edge.rs</strong><dd><code>Update field names in `Artifact` struct for Edge API compatibility</code></dd></summary>
<hr>

rust/src/edge.rs

<li>Updated field names in the <code>Artifact</code> struct to match the new API <br>response.<br> <li> Changed <code>ArtifactName</code> to <code>artifactName</code>.<br> <li> Changed <code>Location</code> to <code>location</code>.<br> <li> Changed <code>Hash</code> to <code>hash</code>.<br> <li> Changed <code>HashAlgorithm</code> to <code>hashAlgorithm</code>.<br> <li> Changed <code>SizeInBytes</code> to <code>sizeInBytes</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14859/files#diff-dcd62c3bd133281f5bd97d8a876bcf753d7a4f3edd6ccbacf44eff3e9fcc484e">+5/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by Fman77** [comment]
What if they switch again to uppercase? Shouldn't we make this case-insensitive?

Is there any way we can make the fix work for previous versions of selenium? Could we contact the MSEdge team and tell them they break Selenium, so that they can un-break it, until we have a newer version that handles case-insensitive attribute names?

**Comment by VietND96** [comment]
Can we make this work for the key name with case-insensitive?

**Comment by Delta456** [comment]
I am not sure if it's a good idea. Plus I don't believe Edge API will break now.

**Comment by VietND96** [comment]
Since via response, all of the keys are not synced in the same standard. Assume one day they are going to change their mind to update the camel case for all other parent keys, we can not guard that.

**Comment by Delta456** [comment]
Do we know all possible options from which they can change into? If yes then maybe we can work upon that.

**Comment by bonigarcia** [comment]
I agree with @Fman77 and @VietND96. I believe the best solution for this issue will be to make this parsing case-insensitive (i.e., accept both `ArtifactName` and `artifactName`, etc.).

@Delta456: Can you please look at the Rust code to make this case insensitive?

Also, I reported this problem to the EdgeWebDriver issue tracker:

https://github.com/MicrosoftEdge/EdgeWebDriver/issues/178

However, the MS Edge WebDriver team is not very responsive to this tracker. It is not the first time I have reported something there (see [here](https://github.com/MicrosoftEdge/EdgeWebDriver/issues/159)), and I didn't have any response, so probably this time will be the same.

**Review by bonigarcia — CHANGES_REQUESTED** [review]
Can we make this case-insentive?

**Comment by Delta456** [comment]
I do not know the best way to make this case insensitive. What preferred way would you suggest?

I also believe the other Struct fields can be renamed with time so we have to make that for all.

**Comment by Delta456** [comment]
> accept both `ArtifactName` and `artifactName`

If we are only going with `ArtifactName` and `artifactName` then we can simply use `#[serde(alias = "...")]`. The following would be the change


```rust
#[derive(Serialize, Deserialize, Debug)]
pub struct Artifact {
    #[serde(rename = "ArtifactName", alias = "artifactName")]
    pub artifact_name: String,
    #[serde(rename = "Location", alias = "location")]
    pub location: String,
    #[serde(rename = "Hash", alias = "hash")]
    pub hash: String,
    #[serde(rename = "HashAlgorithm", alias = "hashAlgorithm")]
    pub hash_algorithm: String,
    #[serde(rename = "SizeInBytes", alias = "sizeInBytes")]
    pub size_in_bytes: u32,
}
```

@bonigarcia what do you think?

**Comment by bonigarcia** [comment]
I've analyzed the required code for this. Indeed, doing this parsing case-insensitive is not straightforward with Rust's serde. In theory, it should be possible to use a custom deserializer and `deserialize_with`. Also, additional existing crates like [serde_aux](https://docs.rs/serde-aux/latest/serde_aux/) do a similar job. But the solution of using an alias is simple and covers the two (in theory) cases we need to face, i.e., `ArtifactName` and `artifactName` (and the other fields). So, I believe it is a good compromise for solving this issue. So please, @Delta456, include those changes in the PR, and I will merge them. Thanks a lot!

**Comment by Delta456** [comment]
@bonigarcia should I also add an alias for `Release` and `EdgeProduct` structs as it can be changed someday soon.

**Comment by bonigarcia** [comment]
I hope that change will not happen in the future. At least, the Microsoft team is now aware that that kind of change can break the clients of their API. But, yes, just in case, we can include a camel-case alias for all the members in those structs - thanks.

**Comment by Delta456** [comment]
Sure, I will do that in a while!

File: rust/src/edge.rs [hunk 1]
```diff
@@ -555,42 +555,42 @@ impl SeleniumManager for EdgeManager {
 
 #[derive(Serialize, Deserialize, Debug)]
 pub struct EdgeProduct {
-    #[serde(rename = "Product")]
+    #[serde(rename = "Product", alias = "product")]
```
Comment (nvborisenko) [inline rust/src/edge.rs:558]: Is it possible to "deserialize case insensitively" globally?

**Review by bonigarcia — APPROVED** [review]

File: rust/src/edge.rs (hunk 1 above)
Comment (bonigarcia) [reply rust/src/edge.rs:558]: I looked for a global solution, but unfortunately, the crate used for serialization (called serde) is not supported out of the box. It is not ideal, but the solution using an alias seems reasonable to me.

**Comment by bonigarcia** [comment]
Thanks, @Delta456!

**Comment by saiganesh98** [comment]
FYI, the older field names are now appearing in the API response, and downloading Edge seems to be working as expected. However, it’s unclear when this change was implemented by the Microsoft team.

**Comment by Delta456** [comment]
Yes, they are making changes without having a changelog or history... 

Glad that we implemented a case-insensitive solution

**Pull request merged**